dashboard/src/app/workspaces/workspace-details/environments/add-machine-dialog/add-machine-dialog.controller.ts

```diff
@@ -0,0 +1,187 @@
+ * Contributors:
+ *   Codenvy, S.A. - initial API and implementation
+ */
+'use strict';
+import {WorkspaceEnvironmentsController} from "../environments.controller";
+
+/**
+ * @ngdoc controller
+ * @name environments.controller:AddMachineDialogController
+ * @description This class is handling the controller for a dialog box about adding a new machine.
+ * @author Oleksii Orel
+ */
+export class AddMachineDialogController {
+  $mdDialog: ng.material.IDialogService;
+  $timeout: ng.ITimeoutService;
+  callbackController: WorkspaceEnvironmentsController;
+  machineRAM: number;
+  machineName: string;
+  environmentKey: string;
+  machineRecipeScript: string;
+  usedMachinesName: Array<string>;
+  machine: Object;
+  machineRecipe: Object;
+  editorOptions: {
+    lineWrapping: boolean,
+    lineNumbers: boolean,
+    matchBrackets: boolean,
+    mode: string,
+    onLoad: Function };
+  environments: Map<any>;
+
+  /**
+   * Default constructor that is using resource
+   * @ngInject for Dependency injection
+   */
+  constructor($mdDialog: ng.material.IDialogService, $timeout: ng.ITimeoutService) {
+    this.$mdDialog = $mdDialog;
+    this.$timeout = $timeout;
+
+    this.usedMachinesName = [];
+    angular.forEach(this.environments, (environment: {machines}, key: string) => {
+      angular.forEach(environment.machines, (machine: Object, machineKey: string) => {
+        this.usedMachinesName.push(machineKey);
+      });
+    });
+
+    this.editorOptions = {
+      lineWrapping: true,
+      lineNumbers: true,
+      matchBrackets: true,
+      mode: 'compose',
+      onLoad: (editor: any) => {
+        this.$timeout(() => {
+          editor.refresh();
+        }, 500);
+      }
+    };
+
+    this.setDefaultData();
+  }
+
+  /**
+   * Set default data.
+   */
+  setDefaultData(): void {
+    this.machine = {
+      'agents': [
+        'org.eclipse.che.terminal', 'org.eclipse.che.ssh'
+      ],
+      'servers': {},
+      'attributes': {}
+    };
+    this.machineRecipeScript = 'new-machine:\n  image: codenvy/ubuntu_jdk8\n  mem_limit: 2147483648\n';
+    this.updateMachineRecipe();
+    if (this.isUnique(this.machineName)) {
+      return;
+    }
+    for (let pos: number = 1; pos < 1000; pos++) {
+      if (this.isUnique(this.machineName + pos.toString())) {
+        this.machineName += pos.toString();
+        break;
+      }
+    }
+    this.updateMachineName(true);
+  }
+
+  /**
+   * Update machine recipe.
+   */
+  updateMachineRecipe(): void {
+    this.machineRecipe = jsyaml.load(this.machineRecipeScript);
+    if (!this.machineRecipe) {
+      return;
+    }
+    let keys: Array<string> = Object.keys(this.machineRecipe);
+    if (keys.length === 0) {
+      return;
+    }
+    this.machineName = keys[0];
+    this.machineRAM = this.machineRecipe[keys[0]]['mem_limit'];
+  }
+
+  /**
+   * Update machine RAM.
+   */
+  updateMachineRAM() {
+    this.machineRecipe[Object.keys(this.machineRecipe)[0]]['mem_limit'] = this.machineRAM;
```

Comment:
use EnvironmentManager when perform any actions or transformations on machine and environment.